src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -0,0 +1,329 @@
+        /// </summary>
+        /// <param name="typeName">A span containing the characters representing the type name to parse.</param>
+        /// <param name="options">An object that describes optional <seealso cref="TypeNameParseOptions"/> parameters to use.</param>
+        /// <returns>Parsed type name.</returns>
+        /// <exception cref="ArgumentException">Provided type name was invalid.</exception>
+        /// <exception cref="InvalidOperationException">Parsing has exceeded the limit set by <seealso cref="TypeNameParseOptions.MaxNodes"/>.</exception>
+        public static TypeName Parse(ReadOnlySpan<char> typeName, TypeNameParseOptions? options = default)
+            => TypeNameParser.Parse(typeName, throwOnError: true, options)!;
+
+        /// <summary>
+        /// Tries to parse a span of characters into a type name.
+        /// </summary>
+        /// <param name="typeName">A span containing the characters representing the type name to parse.</param>
+        /// <param name="options">An object that describes optional <seealso cref="TypeNameParseOptions"/> parameters to use.</param>
+        /// <param name="result">Contains the result when parsing succeeds.</param>
+        /// <returns>true if type name was converted successfully, otherwise, false.</returns>
+        public static bool TryParse(ReadOnlySpan<char> typeName,
+#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB // required by some tools that include this file but don't include the attribute
+            [NotNullWhen(true)]
+#endif
+        out TypeName? result, TypeNameParseOptions? options = default)
+        {
+            result = TypeNameParser.Parse(typeName, throwOnError: false, options);
+            return result is not null;
+        }
+
+        /// <summary>
+        /// Gets the number of dimensions in an array.
+        /// </summary>
+        /// <returns>An integer that contains the number of dimensions in the current type.</returns>
+        /// <exception cref="InvalidOperationException">The current type is not an array.</exception>
+        public int GetArrayRank()
+            => _rankOrModifier switch
+            {
+                TypeNameParserHelpers.SZArray => 1,
+                _ when _rankOrModifier > 0 => _rankOrModifier,
+                _ => throw TypeNameParserHelpers.InvalidOperation_HasToBeArrayClass()
+            };
+
+        /// <summary>
+        /// Returns assembly name which contains this type, or null if this <see cref="TypeName"/> was not
+        /// created from a fully-qualified name.
+        /// </summary>
+        /// <remarks>Since <seealso cref="AssemblyName"/> is mutable, this method returns a copy of it.</remarks>
+        public AssemblyName? GetAssemblyName()
+        {
+            if (_assemblyName is null)
+            {
+                return null;
+            }
+
+#if SYSTEM_PRIVATE_CORELIB
+            return _assemblyName; // no need for a copy in CoreLib
```

Comment:
> no need for a copy in CoreLib

Because...?

Because it's internal?

Reply:
yes, because it's internal and we don't expect to mutate within CoreLib